src/sage/sets/primes.py

```diff
@@ -1,187 +1,851 @@
+            exceptions[x] = False
+        return Primes(self._modulus, self._classes, exceptions)
+
+    def complement_in_primes(self):
+        r"""
+        Return the complement of this set in the set of all prime
+        numbers.
+
+        EXAMPLES::
+
+            sage: P = Primes(modulus=4); P
+            Set of prime numbers congruent to 1 modulo 4: 5, 13, 17, 29, ...
+            sage: Q = P.complement_in_primes(); Q
+            Set of prime numbers congruent to 3 modulo 4 with 2 included: 2, 3, 7, 11, ...
+
+        We check that the union of `P` and `Q` is the whole set of
+        prime numbers::
+
+            sage: P.union(Q)
+            Set of all prime numbers: 2, 3, 5, 7, ...
+
+        and that the intersection is empty::
+
+            sage: P.intersection(Q)
+            Empty set of prime numbers
+
+        .. SEEALSO::
+
+            :meth:`intersection`, :meth:`union`
+        """
+        modulus = self._modulus
+        classes = [c for c in range(modulus)
+                   if c % self._modulus not in self._classes]
+        exceptions = {c: not v for c, v in self._exceptions.items()}
+        return Primes(modulus, classes, exceptions)
+
+    def intersection(self, other):
+        r"""
+        Return the intesection of this set with ``other``.
+
+        INPUT:
+
+        - ``other`` -- a subset of the set of prime numbers
+
+        EXAMPLES::
+
+            sage: P = Primes(modulus=5); P
+            Set of prime numbers congruent to 1 modulo 5: 11, 31, 41, 61, ...
+            sage: Q = Primes(modulus=3, classes=[2]); Q
+            Set of prime numbers congruent to 2 modulo 3: 2, 5, 11, 17, ...
+            sage: P.intersection(Q)
+            Set of prime numbers congruent to 11 modulo 15: 11, 41, 71, 101, ...
+
+        TESTS::
+
+            sage: P.intersection(ZZ) == P
+            True
+            sage: P.intersection(RR)
+            Traceback (most recent call last):
+            ...
+            NotImplementedError: boolean operations are only implemented with other sets of prime numbers
+
+        .. SEEALSO::
+
+            :meth:`complement_in_primes`, :meth:`union`
+        """
+        if other is ZZ:
+            return self
+        if not isinstance(other, Primes):
+            raise NotImplementedError("boolean operations are only implemented with other sets of prime numbers")
+        modulus = self._modulus.lcm(other._modulus)
+        classes = [c for c in range(modulus)
+                   if (c % self._modulus in self._classes
+                   and c % other._modulus in other._classes)]
+        exceptions = {}
+        for c, v in self._exceptions.items():
+            if v and c in other:
+                exceptions[c] = True
+            if not v:
+                exceptions[c] = False
+        for c, v in other._exceptions.items():
+            if v and c in self:
+                exceptions[c] = True
+            if not v:
+                exceptions[c] = False
```

Comment:
Although the following is a bit shorter, I admit that I am not sure whether it is easier to read.  Feel free to ignore.
```suggestion
        exceptions = {c: v for c, v in self._exceptions.items()
                      if not v or c in other}
        exceptions.update((c, v) for c, v in other._exceptions.items()
                          if not v or c in self)
```